docs/pages/prerender-meta.page.server.mdx

```diff
@@ -0,0 +1,73 @@
+import { Link } from '@brillout/docpress'
+
+In some situations we don't want to pre-render a page.
+For example a news page that displays the latest news fetched from a database should be rendered at request-time and certainly not at build-time.
+
+We can use `prerender: false` to tell Vike to skip a page from pre-rendering.
+
+```js
+// /pages/news/+config.h.js
+
+export default {
+  prerender: false
+}
+```
+
+
+## `/renderer/+config.h.js`
+
+We can define `prerender` in higher levels so it applies to more/all pages, e.g. in `/pages/+config.h.js`,
+`/pages/+prerender.js` or even `/renderer/+config.h.js` or `/renderer/+prerender.js`.
+
+For example, to make pre-rendering opt-in instead of opt-out:
+
+```js
+// /renderer/+config.h.js
+
+export default {
+  // By default, our pages are not pre-rendered
+  prerender: false
+}
+```
+
+```js
+// /pages/about-us/+prerender.js
+
+// Opt-in pre-rendering
+export const prerender = true
+```
+
+This is especially handy with <Link text="Domain-Driven File Structure" href='/file-structure#domain-driven' />:
+
+```js
+// /pages/marketing/+prerender.js
+
+// We use Domain-Driven File Structure and `+prerender.js` applies to
+// all our marketing pages.
+
+// Enable pre-rendering for all our marketing pages
+export const prerender = true
+```
+
+
+## Conditional pre-rendering
+
+We can pre-render pages conditionally:
+
+```js
+// /renderer/+prerender.js
+
+export const prerender = isCmsPreview() ? false : true
+```
+
+> This is particularly useful for previewing CMS pages (we don't want to pre-renderer a CMS preview deploy).
+
+Note that <Link text={<><code>onBeforePrerenderStart()</code> hooks</>} href="/onBeforePrerenderStart" /> are always
+called, regardless of `prerender`. If we want to call `onBeforePrerenderStart()` hooks conditionally as well:
+
+{ /* NOTE(aurelien): rewrite this last part after Vike bug fixed where `prerender` seems to have no effect */ }
+
+```js
+export const doNotPrerender = someCondition() ? true : false
+export const prerender = doNotPrerender ? undefined : myPrerenderHook
+```
```

Comment:
@brillout That's the last missing bit - we talked about it offline

Reply:
👍 I'll have a look at it after I'm done with [this feature request](https://github.com/vikejs/vike/issues/1231) (which is a blocker for [a high-tier sponsor](https://github.com/Alignable)).

Reply:
I just checked https://vike.dev/doNotPrerender#conditional-pre-rendering and the use case described there (disabling pre-rendering for "deploy live previews") should, AFAICT, also work with the V1 design.

The idea being to set an env `DISABLE_PRERENDER` and then setting `prerender: !process.env.DISABLE_PRERENDER` and `onBeforePrerenderStart: process.env.DISABLE_PRERENDER ? null : onBeforePrerenderStart`.

Reply:
indeed it's easier to do

```js
export const doNotPrerender = someCondition() ? true : false
export const prerender = someCondition() ? undefined : myPrerenderHook
```

that

```js
export const doNotPrerender = someCondition() ? true : false
export const prerender = doNotPrerender ? undefined : myPrerenderHook
```

the second one would probably require an effect, and would be quite hard to describe in the documentation. I'll go for the first one then.

Reply:
So this doesn't work:

```js
import { onBeforePrerenderStart } from './myOnBeforePrerenderStartHook.js'
import type { Config } from 'vike/types'

export default {
  prerender: someCondition(),
  onBeforePrerenderStart: someCondition() ? onBeforePrerenderStart : undefined
} satisfies Config

function someCondition() {
  return false
}
```

if the condition is `false`, of course we get

```
Error: [vike][Wrong Usage] /pages/about/+config.h.ts imports the following:
  import { onBeforePrerenderStart } from "./myOnBeforePrerenderStartHook.js";
But the import onBeforePrerenderStart isn't re-exported at export default { ... } and therefore has no effect, see explanation at https://vike.dev/header-file
```

That's understandable but a bit of a pity. This will have to be solved with a `+onBeforePrerenderStart.js` file. Giving it a try

Reply:
I think we can remove that assertUsage(). The trade off here is that we lose a bit on educating users about config.h.js but it could be worth it 🤔.

Or we just turn it into a warning. At least it isn’t blocking then.

Reply:
There seems to be a barrier in every path I take:

```js
// +onBeforePrerenderStart.js

export { onBeforePrerenderStart }

import { someCondition } from './someCondition'

const onBeforePrerenderStart = someCondition()
  ? async () => {
      return ['/about']
    }
  : null

// fails with
// Error: [vike][Wrong Usage] Set onBeforePrerenderStart to null in a +config.h.js file instead of /pages/about/+onBeforePrerenderStart.js
```

```js
// +onBeforePrerenderStart.js

export { onBeforePrerenderStart }

import { someCondition } from './someCondition'

const onBeforePrerenderStart = someCondition()
  ? async () => {
      return ['/about']
    }
  : undefined

// fails with
// Error: [vike][Wrong Usage] Hook onBeforePrerenderStart() defined by /pages/about/+onBeforePrerenderStart.js should be a function
```

I don't think there is a working V1-way for unsetting a hook based on a condition at the moment. Am I missing something?

Reply:
maybe the assertUsage that could be lifted, because not-dangerous, is the one preventing me from exporting `null` in `+onBeforePrerenderStart.js`?

Reply:
> // Error: [vike][Wrong Usage] Set onBeforePrerenderStart to null in a +config.h.js file instead of /pages/about/+onBeforePrerenderStart.js

This one is necessary because Vike needs to know whether a global hook has been suppressed (i.e. set to `null`). (Vike doesn't load hooks at config-time, so the only way to tell Vike that a hook is `null` at config-time is to set it in `+config.h.js`.)

<hr>

How about we do the following? It seems to be the quickest way to unblock us. Thoughts?

> Or we just turn it into a warning. At least it isn’t blocking then.

It'll give us some more time until users complain about the warning. (And I think it's okay-ish if we don't tell users about the warning in the docs if we feel lazy.)